compaction/compaction_strategy.cc

```diff
@@ -176,13 +176,30 @@ double size_tiered_backlog_tracker::backlog(const compaction_backlog_tracker::on
     return b > 0 ? b : 0;
 }
 
-// FIXME: Should provide strong exception safety guarantees
+// Provides strong exception safety guarantees.
 void size_tiered_backlog_tracker::replace_sstables(const std::vector<sstables::shared_sstable>& old_ssts, const std::vector<sstables::shared_sstable>& new_ssts) {
+    std::vector<sstables::shared_sstable> undo_erase;
+    std::vector<sstables::shared_sstable> undo_insert;
+    undo_erase.reserve(old_ssts.size());
+    undo_insert.reserve(new_ssts.size());
+    auto undo = defer([&] () noexcept {
+        clogger.warn("size_tiered_backlog_tracker: replace_sstables failed. Rolling back...");
+        for (auto& sst : undo_insert) {
+            _total_bytes -= sst->data_size();
+            _all.erase(std::move(sst));
+        }
+        for (auto& sst : undo_erase) {
+            _total_bytes += sst->data_size();
+            _all.insert(std::move(sst));
```

Comment:
This certainly allocates.

Reply:
since this is in undo, mu assumption is that it is not likely to throw since those elements were there before they were erased, so it's similar to inserting after reserving capacity.

Reply:
Alternatively, we can copy `_all` and restore it on error.
This is simpler and more robust, yet less efficient.

Reply:
I don't think we can make that assumption.

What we can do is, during the initial operation, only erase after all insertions are done. Then, if erases don't throw, then once insertions are done we won't need to roll back, and therefore won't need to re-insert erased elements.

There's still the problem that interval_map erase needs to allocate.

Reply:
> There's still the problem that interval_map erase needs to allocate.

But what prompted you about it?
I don't see it used in any of the backlog trackers.

Are you thinking about `partitioned_sstable_set`?
We're not there yet...

Reply:
> What we can do is, during the initial operation, only erase after all insertions are done. Then, if erases don't throw, then once insertions are done we won't need to roll back, and therefore won't need to re-insert erased elements.

Good idea.  I'mm make that change.

Reply:
std::unordered_set::erase may allocate too during hashing, but I don't think it happens with lw_shared_ptr, which hashing and comparison are done on underlying ptr.

I think the simplest way is to copy _all and restore on failure. STCS doesn't have a high number of sstables anyway.

Reply:
Isn't sstable_set manipulated here?

Reply:
It's a bit more complicated since `refresh_sstables_backlog_contribution` may still throw (e.g. from `get_buckets` -> `create_sstable_and_length_pairs`) and we'll need to undo the erasures in this case.
I think I'll go back to the calculate_sstables_backlog_contribution that will do the calculation without changing the tracker state, based on a provisional list which has `_all + new_ssts - old_ssts` and then applying the resulting `sstables_backlog_contribution` object will be noexcept and can be done only when all else succeeded.

Reply:
> Isn't sstable_set manipulated here?

the tracker is essentially storing a copy of the main sstable set, so it must reflect what's stored in the table. as having two sources of truth is asking for trouble, I want the tracker to be feeded the current set, rather than having the tracker maintaining the set by itself.

Reply:
> It's a bit more complicated since `refresh_sstables_backlog_contribution` may still throw (e.g. from `get_buckets` -> `create_sstable_and_length_pairs`) and we'll need to undo the erasures in this case. I think I'll go back to the calculate_sstables_backlog_contribution that will do the calculation without changing the tracker state, based on a provisional list which has `_all + new_ssts - old_ssts` and then applying the resulting `sstables_backlog_contribution` object will be noexcept and can be done only when all else succeeded.

Now that I put more thought on it, I think we can't really allow the set maintained by the tracker to deviate from the table's sstable set. Otherwise, future calculations will all be done on incorrect / outdated sstable set. Probably one of the reasons that Glauber just disabled the tracker on exception. 

We could store pending updates in the tracker and retry them in the future (but pushing the missing updates into a list can fail too), but that's probably too complicated for this.

Reply:
Maybe the easiest way is to just kill the set maintained by the tracker.

Reply:
simplified in v4 by applying all changes to temporary variables instead of in place.

Reply:
> Maybe the easiest way is to just kill the set maintained by the tracker.

It's not that easy since stcs tracker is also embedded in twcs and lcs trackers, that will need to maintain that external set for stcs tracker.

I'll look into it more deeply as a followup.